statsmodels/robust/covariance.py

```diff
@@ -0,0 +1,607 @@
+High-Dimensional Covariance Matrices.”
+IEEE Transactions on Signal Processing 59 (9): 4097–4107.
+doi:10.1109/TSP.2011.2138698.
+
+
+"""
+
+import numpy as np
+from scipy import linalg, stats
+from statsmodels import robust
+import statsmodels.robust as robust
+
+# shorthand functions
+mad = robust.mad
+mad0 = lambda x: mad(x, center=0)
+
+class Holder(object):
+    def __init__(self, **kwds):
+        self.__dict__.update(kwds)
+
+
+### GK and OGK ###
+
+def _weight_mean(x, c):
+    x = np.asarray(x)
+    w = (1 - (x / c)**2)**2 * (np.abs(x) <= c)
+    return w
+
+def _winsor(x, c):
+    return np.minimum(x**2, c**2)
+
+
+def scale_tau(data, cm=4.5, cs=3, weight_mean=_weight_mean,
+              weight_scale=_winsor, normalize=True, ddof=2):
+    """tau estimator of univariate scale
+
+    Experimental, API will change
+
+    Parameters
+    ----------
+    data : array_like, 1-D or 2-D
+        If data is 2d, then the location and scale estimates
+        are calculated for each column
+    cm : float
+        constant used in call to weight_mean
+    cs : float
+        constant used in call to weight_scale
+    weight_mean : callable
+        function to calculate weights for weighted mean
+    weight_scale : callable
+        function to calculate scale, "rho" function
+    normalize : bool
+        rescale the scale estimate so it is consistent when the data is
+        normally distributed. The computation assumes winsorized (truncated)
+        variance.
+
+    Returns
+    -------
+    mean : nd_array
+        robust mean
+    std : nd_array
+        robust estimate of scale (standard deviation)
+
+    Notes
+    -----
+    Uses definition of Maronna and Zamar 2002, with weighted mean and
+    trimmed variance.
+    The normalization has been added to match R robustbase.
+    R robustbase uses by default ddof=0, with option to set it to 2.
+
+    """
+
+    x = np.asarray(data)
+    nobs = x.shape[0]
+
+    med_x = np.median(x, 0)
+    xdm = x - med_x
+    mad_x = np.median(np.abs(xdm), 0)
+    wm = weight_mean(xdm / mad_x, cm)
+    mean = (wm * x).sum(0) / wm.sum(0)
+    var = mad_x**2 * weight_scale((x - mean) / mad_x, cs).sum(0) / (nobs - ddof)
+
+    cf = 1
+    if normalize:
+        c =  cs * stats.norm.ppf(0.75)
+        cf = 2 * ((1 - c**2) * stats.norm.cdf(c) - c * stats.norm.pdf(c)
+                 + c**2) - 1
+    #return Holder(loc=mean, scale=np.sqrt(var / cf))
+    return mean, np.sqrt(var / cf)
+
+
+
+def mahalanobis(data, cov=None, cov_inv=None):
+    """Mahalanobis distance
+
+    """
+    x = np.asarray(data)
+    if cov_inv is not None:
+        d = (x * cov_inv.dot(x)).sum(1)
+    elif cov is not None:
+        d = (x * np.linalg.solve(cov, x.T).T).sum(1)
+    else:
+        raise ValueError('either cov or cov_inv needs to be given')
+
+    return d
+
+
+def cov_gk1(x, y, scale_func=mad):
+    """Gnanadesikan and Kettenring covariance between two variables
+
```

Comment:
maybe add option for general version, i.e. a, b coefficient as in Huber/Ronchetti book (?) 
or normalized by scale of x and of y.